bot.go

```diff
@@ -1,36 +1,379 @@
+
+func defaultSendOpt() *tba.SendOptions {
+
+	return &tba.SendOptions{}
+}
+
+func ensureDataFileExists(fileName string) {
+
+	if _, err := os.Stat(fileName); err != nil {
+
+		err = ioutil.WriteFile(fileName, []byte("{}"), 0644)
 		if err != nil {
-			log.Fatal("error sending version")
+
+			log.Fatalf("could not create %s", fileName)
+		}
+	}
+}
+
+func output(b *Bot) {
+
+	for {
+
+		select {
+
+		case m := <-b.output:
+			sendMessage(b, m.chatID, m.text, defaultSendOpt())
+		}
+	}
+}
+
+func inputKiller(c chan int64, m map[int64]chan string) {
+	for {
+
+		select {
+
+		case id := <-c:
+			delete(m, id)
+		}
+	}
+}
+func sendMessage(b *Bot, chat int64, message string, sendOpt *tba.SendOptions) {
+
+	tmpChat := tba.Chat{ID: chat, Title: "", FirstName: "", LastName: "", Type: "", Username: ""}
+	_, err := b.api.Send(&tmpChat, message, sendOpt)
+
+	if err != nil {
+
+		log.Printf("Could not send message %s to %d", message, chat)
+	}
+}
+
+func getAnswer(in chan string) (string, error) {
+
+	timeout := 0
+	for timeout < 1200 {
+
+		select {
+
+		case a := <-in:
+			return a, nil
+		default:
+			time.Sleep(250 * time.Millisecond)
+			timeout++
+		}
+	}
+
+	return "", errors.New("timeout")
+}
+
+func dialog(out chan msg, chatID int64, question string, in chan string) (string, error) {
+
+	out <- msg{chatID, question}
+	a, err := getAnswer(in)
+	if err != nil {
+
+		out <- msg{chatID, "Przekroczono czas odpowiedzi"}
+	}
+	return a, err
+
+}
+
+func addFlashcard(flashcards map[int64]map[string]map[string]string, chatID int64, out chan msg, in chan string, state chan int64) {
+
+	topic, err := dialog(out, chatID, "Podaj temat", in)
+	if err != nil {
+		state <- chatID
+		return
+	}
+
+	term, err := dialog(out, chatID, "Podaj pojecie", in)
+	if err != nil {
+		state <- chatID
+		return
+	}
+
+	if _, ok := flashcards[chatID][topic][term]; ok {
+
+		out <- msg{chatID, "Fiszka juz istnieje, edytuj za pomoca /edytujfiszke"}
+		state <- chatID
+		return
+	}
+
+	definition, err := dialog(out, chatID, "Podaj definicje", in)
+	if err != nil {
+		state <- chatID
+		return
+	}
+
+	if flashcards[chatID] == nil {
+
+		flashcards[chatID] = make(map[string]map[string]string)
+	}
+
+	if flashcards[chatID][topic] == nil {
+
+		flashcards[chatID][topic] = make(map[string]string)
+	}
+
+	flashcards[chatID][topic][term] = definition
+
+	flashcardsJSON, err := json.Marshal(flashcards)
+	if err != nil {
+		log.Print("Could not encode flashcards")
+		out <- msg{chatID, "Wystapil problem, sprobuj pozniej"}
+		state <- chatID
+		return
+	}
+
+	err = ioutil.WriteFile(flashcardsFileName, flashcardsJSON, 0644)
+	if err != nil {
+		log.Print("Could not write flashcards")
+		out <- msg{chatID, "Wystapil problem, sprobuj pozniej"}
+		state <- chatID
+		return
+	}
+
+	out <- msg{chatID, "Dodano fiszke"}
+	state <- chatID
+}
+
+func displayFlashcard(flashcards map[string]map[string]string, m *tba.Message, output chan msg) {
+
+	chatID := m.Chat.ID
+
+	if m.Text == "/fiszka" {
```

Comment:
A jak to zadziała jak dasz spację na końcu? Są usuwane?

https://golang.org/pkg/strings/#Trim